### PR TITLE
[ci] Drop support for Python 3.7 and Python 3.8

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

In our CI, we are still testing for old Python versions, namely Python 3.7 and Python 3.8. However:

- Python 3.7 and Python 3.8 have both reached end of life (see https://devguide.python.org/versions/)
- Almost no other package still supports Python 3.7 and 3.8 (XGBoost v3.0.0 released a couple weeks ago has even dropped support for Python 3.9)
- `conda-forge` does not build packages for Python <3.9 anymore

I'd argue that we can simplify our CI by dropping support for 3.7 and 3.8 and, instead, test on 3.9 the new "oldest supported Python version".

# Changes

- Adjust relevant CI jobs to run for Python 3.9 instead of 3.8 (and 3.7)

### CODE REVIEW

**Comment by borchero** [comment]
@jameslamb the AppVeyor CI fails to solve the environment. However, I can't replicate this behavior locally (even for Windows) -- any idea how to debug this?

**Comment by jameslamb** [comment]
> the AppVeyor CI fails to solve the environment.

Links and a bit of plaintext logs would be helpful, to narrow it down. "fails to solve the environment" could take many different forms.

I just looked in the history of Appveyor runs 
 and found this one: https://ci.appveyor.com/project/guolinke/lightgbm/builds/51955916

It looks like the solve setting up the environment succeeded **as did all of the Python unit tests**.

```text
...
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
#
# To activate this environment, use
#
#     $ conda activate test-env
...
==== 878 passed, 33 skipped, 4 xfailed, 100 warnings in 89.26s (0:01:29) =====
```

I guess you probably mean this later solve, after the tests:

```text
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... 
```

Since it's after the tests, I guess it must be this one:

https://github.com/microsoft/LightGBM/blob/ec5492f8f915f9bab99e52417cff6808dbdb8ddd/.ci/test-windows.ps1#L146

> any idea how to debug this?

Do this initial `conda` env creation locally on windows:

https://github.com/microsoft/LightGBM/blob/ec5492f8f915f9bab99e52417cff6808dbdb8ddd/.ci/test-windows.ps1#L79-L85

Run `conda env export --name test-env` and compare the package versions to what you see in the Appveyor logs.

If there are differences in any package versions, then that means there's something different between your local environment and CI, and those differences are the result of the different behavior. The relevant difference will probably exist in the output of these commands:

```shell
conda config --get
conda info
conda --version
```

If all of the versions and `conda info` stuff are identical, then you could start investigating other, less likely, root causes like available memory / disk, network bandwidth, environment variables, etc.

**Comment by borchero** [comment]
Thanks for the hint @jameslamb, `tornado` caused some issues :)

File: .github/workflows/python_package.yml [hunk 1]
```diff
@@ -75,32 +75,32 @@ jobs:
         include:
           - os: macos-13
             task: regular
-            python_version: '3.11'
+            python_version: "3.11"
```
Comment (jameslamb) [inline .github/workflows/python_package.yml:78]: Why were all these quoting changes necessary? If they were just something done automatically by your IDE, please revert them.

If there's a way to enforce a specific choice of quote type with `yamllint`, I'd welcome a separate PR to do that, so we don't have to think about this in the future.

**Review by jameslamb — DISMISSED** [review]
Thanks for this! I left one small suggestion for your consideration.

Overall, I'm supporting of changing the Python support policy in this project to **Python version support is dropped when a version of Python reaches end-of-life, according to https://devguide.python.org/versions/**.

That'd be more aggressive than we've been in the past, but I think it's defensible given that many of `lightgbm`'s dependencies are even more aggressive ([SPEC 0](https://scientific-python.org/specs/spec-0000/)) and so is `conda-forge` (https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-python.html#managing-python).

But let's not merge this until @StrikerRUS has a chance to comment as well, I'd like to hear his thoughts.

---

Also... @borchero , I know it's been a little while since your last PR here. As a reminder... every PR in this repo needs to have a label from the set of release notes sections:

https://github.com/microsoft/LightGBM/blob/ec5492f8f915f9bab99e52417cff6808dbdb8ddd/.github/release-drafter.yml#L3-L15

Please add them on all of the PRs you've opened. This one would be `breaking`, similar to when we dropped Python 3.6: #6459

File: .ci/conda-envs/ci-core-py39.txt [hunk 2]
```diff
@@ -10,27 +10,24 @@
...
 cffi=1.15.*
-# older versions of Dask are incompatible with pandas>=2.0, but not all conda packages' metadata accurately reflects that
-#
-# ref: https://github.com/microsoft/LightGBM/issues/6030
-dask=2022.2.*
-distributed=2022.2.*
+# Dask & distributed are the first versions that support tornado >=6.2
+# which is required for testing
+dask=2022.12.*
+distributed=2022.12.*
```
Comment (jameslamb) [inline .ci/conda-envs/ci-core-py39.txt:24]: I'd support updating the minimum for `dask` in the `[dask]` extra to `>=2022.12.0` to match this.

Here:

https://github.com/microsoft/LightGBM/blob/ec5492f8f915f9bab99e52417cff6808dbdb8ddd/python-package/pyproject.toml#L42

That's 2+ years old at this point, which is a LOT of time given the speed of development in Dask.

**Comment by StrikerRUS** [comment]
I don't mind to change the Python support policy. I believe this change will be understandable from the community point of view given our small team size.

Also, nowadays there are a lot of [cool] environment managing tools besides conda that extremely simplify the process of creating new envs. So for newcomers it should be a problem. And for old users, as @jameslamb already mentioned, they are locked with old package versions including the deps.

P.S. Another example of harsh policy for version retirement: #6520.

**Review by StrikerRUS — APPROVED** [review]
LGTM!

File: .ci/conda-envs/ci-core-py39.txt (hunk 2 above)
Comment (jameslamb) [reply .ci/conda-envs/ci-core-py39.txt:24]: I'll do this in a follow-up PR, let's not grow the size of this PR too much more.

File: .github/workflows/python_package.yml (hunk 1 above)
Comment (jameslamb) [reply .github/workflows/python_package.yml]: I removed these unrelated changes in https://github.com/microsoft/LightGBM/pull/6896/commits/8c9bec8a2ed164d5fda5076be6ddf42d0d018608

**Comment by jameslamb** [comment]
It's been 3 months without a response from @borchero (who mentioned to me offline that he has not had much time for LightGBM recently), so I took this over.

I just pushed the following changes:

* merged in latest `master`
* removed all the unnecessary quoting changes (https://github.com/microsoft/LightGBM/pull/6896#discussion_r2063846170)
* added a short section to the Python docs documenting the new Python support policy
* clarified a comment about `dask` / `distributed` versions

Since I pushed changes here, my review shouldn't count towards a merge.

@StrikerRUS if you don't mind, could you please re-review?

**Review by StrikerRUS — APPROVED** [review]
@jameslamb Thanks a lot for picking this up!
